
学习图像上的算术运算 加法 减法 位运算等
## 1 图像的加法运算

### 1.1 图像相加 add() 
2个图像对象img，img2如果通道数一样、大小一样，可以使用cv2.add(img,img2)相加，得到一个新的图像。 

In [ ]:
import cv2
import numpy as np

img1 = cv2.imread('./images/lena.jpg')
img2 = cv2.imread('./images/opencv_logo.png')

dst = img1+img2

In [6]:
img1 = cv2.imread('./images/lena.jpg')[:512,:512]
img2 = cv2.imread('./images/opencv_logo.png')[:512,:512]

#dst = img1+img2
dst = cv2.add(img2,img1)
cv2.imshow('image add ', dst)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [7]:
img1[111,222],img2[111,222],dst[111,222]

(array([146, 151, 206], dtype=uint8),
 array([  0,   0, 255], dtype=uint8),
 array([146, 151, 255], dtype=uint8))

2个图像中白色部分会覆盖了原来的位置，那是因为add()方法相加的图像如果超过了阈值会被截断。 

在OpenCV中图像的算术运算遵循“饱和运算”的规则，如果计算的结果超过了阈值范围，则就近进行截断，比如unit8类型的数据范围是【0,255】，如果2个数值直接相加的结果大于255就会赋值为255。 

### 1.2 图像相加 运算符+

In [2]:

x = np.uint8([250])
y = np.uint8([10])


In [3]:
cv2.add(x, y) # 250+10 = 260 => 255

array([[255]], dtype=uint8)

In [4]:
x+y # 250+10=260%256=4

array([4], dtype=uint8)

In [ ]:
img1 = cv2.imread('./images/lena.jpg')[:512,:512]
img2 = cv2.imread('./images/opencv_logo.png')[:512,:512]

dst = img1+img2
cv2.imshow('image add ', dst)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [9]:
img1[111,222],img2[111,222],dst[111,222]

(array([146, 151, 206], dtype=uint8),
 array([  0,   0, 255], dtype=uint8),
 array([146, 151, 205], dtype=uint8))

从运行结果看，+运算符将2个图像相加，新图像和lena.jpg差异不大，但是和opencv-logo.png相差甚远。

从像素[111,222]的运算结果看，当2个值相加大于255时，会将这个值对256求余得到新的像素值。  

为什么用符号+运算后的结果会对256求模？ 
- numpy数组的数据类型是unit8, C语言中unsigned char类型的数据表示的范围是[0,255]的闭区间


### 1.3 加权加法 addWeighted()
addWighted()可以用来实现2幅图像的不同权值的加法，实现图像混合的效果。 


函数cv2.addWeighted()可以按下面的公式对图片进行混合操作。

$g(x)=(1-a)f_0(x)+af_1(x)+b$


现在我们把两幅图混合在一起。第一幅图的权重是0.7,第二幅图的权重是0.3, b=0。



In [10]:
img2_dst = cv2.resize(img2,(img1.shape[0],img1.shape[1])) #放大图像

dst = cv2.addWeighted(img1, 0.7, img2_dst, 0.3, 0)  # 第一幅图的权重是 0.7 第二幅图的权重是 0.3

cv2.imshow('dst', dst)
cv2.waitKey(0) 
cv2.destroyAllWindows()

In [11]:
img1[111,222],img2[111,222],dst[111,222]

(array([146, 151, 206], dtype=uint8),
 array([  0,   0, 255], dtype=uint8),
 array([102, 106, 221], dtype=uint8))

In [10]:

import cv2
import numpy as np


img1=cv2.imread('./images/subtract1.jpg',1)#灰度图
img2=cv2.imread('./images/subtract2.jpg',1)

cv2.imshow('subtract1',img1)
cv2.imshow('subtract2',img2)

#
st=img2-img1
# st=img1-img2#相反
cv2.imshow('after subtract',st)

#效果好一点
# ret,threshold=cv2.threshold(st,0, 127, cv2.THRESH_BINARY)
ret,threshold=cv2.threshold(st, 50,255, cv2.THRESH_BINARY)
cv2.imshow('after threshold', threshold)


cv2.waitKey(0)
cv2.destroyAllWindows()

## 2 图像减法运算 

减法运算的subtract()也遵循“饱和运算”规则，如果2个图像相减直接计算结果小于0，也会被截断到255。 

### 2.1 图像相减 subtract() 

2个图像相减仍然要求通道数一样，图像尺寸一样。

In [15]:
img1 = cv2.imread('./images/lena.jpg')[:512,:512]
img2 = cv2.imread('./images/opencv_logo.png')[:512,:512]

dst1 = cv2.subtract(img1,img2) 
dst2 = cv2.subtract(img2,img1)
cv2.imshow('image subtract1', dst1)
cv2.imshow('image subtract2', dst2)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [18]:
img1[111,222],img2[111,222],dst1[111,222],dst2[111,222]

(array([146, 151, 206], dtype=uint8),
 array([  0,   0, 255], dtype=uint8),
 array([146, 151, 207], dtype=uint8),
 array([110, 105,  49], dtype=uint8))

### 2.2 图像相减 运算符-

In [17]:
img1 = cv2.imread('./images/lena.jpg')[:512,:512]
img2 = cv2.imread('./images/opencv_logo.png')[:512,:512]

dst1 = img1-img2
dst2 = img2-img1
cv2.imshow('img1-img2', dst1)
cv2.imshow('img2-img1', dst2)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [20]:
img1[111,222],img2[111,222],dst1[111,222],dst2[111,222]

(array([146, 151, 206], dtype=uint8),
 array([  0,   0, 255], dtype=uint8),
 array([146, 151, 207], dtype=uint8),
 array([110, 105,  49], dtype=uint8))

### 2.3 绝对值减法 absdiff() 

而不是像subtract()和符号-减法那样做饱和运算或者求模，absdiff()得到的是2个图像间像素的绝对差值。

In [22]:
img1 = cv2.imread('./images/lena.jpg')[:512,:512]
img2 = cv2.imread('./images/opencv_logo.png')[:512,:512]

dst1 = cv2.absdiff(img1,img2)
dst2 = cv2.absdiff(img2,img1)
cv2.imshow('absdiff1', dst1)
cv2.imshow('absdiff2', dst2)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [23]:
img1[111,222],img2[111,222],dst1[111,222],dst2[111,222]

(array([146, 151, 206], dtype=uint8),
 array([  0,   0, 255], dtype=uint8),
 array([146, 151,  49], dtype=uint8),
 array([146, 151,  49], dtype=uint8))

### 2.4 图像和标量加减  

图像和标量进行加减，则是将图像的每一个元素都和这个标量值进行加减。 

如果要进行多通道运算，标量值则使用一个包含4个数值的元组表示。

In [28]:
img1 = cv2.imread('./images/lena.jpg')[:512,:512]

dst1 = cv2.add(img1,50)
dst2 = cv2.add(img1,(50,50,50,50))

cv2.imshow('add(img,50)', dst1)
cv2.imshow('add(img,(50,50,50,0))', dst2)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [27]:
img1[111,222],dst1[111,222],dst2[111,222]

(array([146, 151, 206], dtype=uint8),
 array([196, 151, 206], dtype=uint8),
 array([196, 201, 255], dtype=uint8))

## 3 图像乘法  
### 3.1 图像乘法multiply()  

multiply()用法：dst=cv2.multiply(src1, src2[, dst[, scale[, dtype]]])，src1和src2为图像对象，可选参数scale为放大倍数。

结果dst=saturate(scale*src1*src2)

multiply()遵循饱和运算规则，比如uint8类型的数据如果超过255会被截断到255

In [34]:
img1 = cv2.imread('./images/lena.jpg')[:512,:512]
img2 = cv2.imread('./images/opencv_logo.png')[:512,:512]

dst1 = cv2.multiply(img1,img2)
cv2.imshow('multiply', dst1)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [39]:
img1[111,222],img2[111,222],dst1[111,222]

(array([146, 151, 206], dtype=uint8),
 array([  0,   0, 255], dtype=uint8),
 array([ 0,  0, 50], dtype=uint8))

从图像学意义上看，因为opencv-logo.png的每个像素要么是0要么是255

用multiply()计算它和其他的图像相乘得到的结果也是0或者255，所以最后的图像和opencv-logo.png一样。

### 3.2  符号乘法

符号乘法实际就是numpy数组的乘法，和用+/-做numpy加减法一样，在数值类型表示范围的上限加1取模，比如uint8类型的数据对256取模。

In [36]:
img1 = cv2.imread('./images/lena.jpg')[:512,:512]
img2 = cv2.imread('./images/opencv_logo.png')[:512,:512]

dst1 = img1*img2
cv2.imshow('multiply', dst1)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [38]:
img1[111,222],dst2[111,222],dst1[111,222]

(array([146, 151, 206], dtype=uint8),
 array([196, 201, 255], dtype=uint8),
 array([ 0,  0, 50], dtype=uint8))

## 4 图像除法  

### 4.1 图像除法divide()  

divide()有2种用法：

dst = cv2.divide( src1, src2[, dst[, scale[, dtype]]] )：第1个和第2个位置参数都是图像对象，可选参数scale参数指定src1的放大倍数


dst = cv2.divide( scale, src2[, dst[, dtype]] )：第1个位置参数为数值类型，第2个位置参数为图像对象

如果是uint8等整数类型的除法，运算后的结果会做四舍五入取整。divide()除法也遵守“饱和运算规则” 

In [40]:
img1 = cv2.imread('./images/lena.jpg')[:512,:512]
img2 = cv2.imread('./images/opencv_logo.png')[:512,:512]

dst1 =cv2.divide(img1,img2)
dst2 =cv2.divide(img2,img1)
cv2.imshow('divide 1 ', dst1)
cv2.imshow('divide 2 ', dst2)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [42]:
img1[111,222],img2[111,222],dst2[111,222],dst1[111,222]

(array([146, 151, 206], dtype=uint8),
 array([  0,   0, 255], dtype=uint8),
 array([0, 0, 1], dtype=uint8),
 array([0, 0, 1], dtype=uint8))

### 4.2  图像除法 / 
符号除法实际就是numpy数组的除法。 

In [43]:
img1 = cv2.imread('./images/lena.jpg')[:512,:512]
img2 = cv2.imread('./images/opencv_logo.png')[:512,:512]

dst1 =img1/img2
dst2 =img2/img1
cv2.imshow('img1/img2 ', dst1)
cv2.imshow('img2/img1 ', dst2)
cv2.waitKey(0)
cv2.destroyAllWindows()

<ipython-input-43-90dd9eb9d555>:4: RuntimeWarning: divide by zero encountered in divide
  dst1 =img1/img2
<ipython-input-43-90dd9eb9d555>:4: RuntimeWarning: invalid value encountered in divide
  dst1 =img1/img2
<ipython-input-43-90dd9eb9d555>:5: RuntimeWarning: divide by zero encountered in divide
  dst2 =img2/img1
<ipython-input-43-90dd9eb9d555>:5: RuntimeWarning: invalid value encountered in divide
  dst2 =img2/img1


In [45]:
dst1.dtype

dtype('float64')

imread()读入图像默认的数据类型为uint8，符号除法得到的数据类型发生了改变，变成了float64。

当除数为0时，numpy除法打印了告警信息：divide by zero encountered in true_divide

直接计算结果为inf 

当有元素为0且作为被除数时，divide()计算仍然是有实际意义的

## 5 图像的位运算  

图像的位运算是指对图像的数值按照二进制值逐位进行取反、与、或、异或操作。

### 5.1 按位取反bitwise_not()  
按位取反就是将数值根据每个bit位1变0，0变1，比如0xf0按位取反就变成了0x0f，如果是uint8类型的数据，取反前后的数据相加结果为0xff(255) 

In [69]:
hex(255),hex(0),int("0xf0",16),int("0x0f",16)

('0xff', '0x0', 240, 15)

In [72]:
bin(int("0xf0",16)),(~int("0xf0",16)+256)

('0b11110000', 15)

In [73]:
img1 = cv2.imread('./images/lena.jpg')[:512,:512]
img2 = cv2.imread('./images/opencv_logo.png')[:512,:512]

dst1 =cv2.bitwise_not(img1)
dst2 =cv2.bitwise_not(img2)
cv2.imshow('lena-bitwise-not ', dst1)
cv2.imshow('opencv-bitwise-not ', dst2)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [75]:
img1[111,222],img2[111,222],dst1[111,222],dst2[111,222]

(array([146, 151, 206], dtype=uint8),
 array([  0,   0, 255], dtype=uint8),
 array([109, 104,  49], dtype=uint8),
 array([255, 255,   0], dtype=uint8))

### 5.2 按位与bitwise_and()、或bitwise_or()、异或bitwise_xor() 

2个图像的按位操作和算术运算一样，也要求2个图像的大小一样，通道数一样。 

In [76]:
img1 = cv2.imread('./images/lena.jpg')[:512,:512]
img2 = cv2.imread('./images/opencv_logo.png')[:512,:512]

dst1 =cv2.bitwise_and(img1,img2)
dst2 =cv2.bitwise_or(img1,img2)
dst3 =cv2.bitwise_xor(img1,img2)
cv2.imshow('lena-bitwise_and ', dst1)
cv2.imshow('lena-bitwise_or ', dst2)
cv2.imshow('lena-bitwise_xor ', dst3)
cv2.waitKey(0)
cv2.destroyAllWindows()